crates/crates_io_database/src/models/email.rs

```diff
@@ -24,46 +25,65 @@ pub struct NewEmail<'a> {
+    #[builder(default = false)]
+    pub primary: bool,
 }
 
 impl NewEmail<'_> {
-    pub async fn insert(&self, conn: &mut AsyncPgConnection) -> QueryResult<()> {
+    pub async fn insert(&self, conn: &mut AsyncPgConnection) -> QueryResult<Email> {
         diesel::insert_into(emails::table)
             .values(self)
-            .execute(conn)
-            .await?;
-
-        Ok(())
+            .returning(Email::as_returning())
+            .get_result(conn)
+            .await
     }
 
-    /// Inserts the email into the database and returns the confirmation token,
-    /// or does nothing if it already exists and returns `None`.
-    pub async fn insert_if_missing(
+    /// Inserts the email into the database and returns it, unless the user already has a
+    /// primary email, in which case it will do nothing and return `None`.
+    pub async fn insert_primary_if_missing(
         &self,
         conn: &mut AsyncPgConnection,
-    ) -> QueryResult<Option<SecretString>> {
-        diesel::insert_into(emails::table)
-            .values(self)
-            .on_conflict_do_nothing()
-            .returning(emails::token)
-            .get_result::<String>(conn)
-            .await
-            .map(Into::into)
-            .optional()
+    ) -> QueryResult<Option<Email>> {
+        // Check if the user already has a primary email
+        let primary_count = emails::table
+            .filter(emails::user_id.eq(self.user_id))
+            .filter(emails::primary.eq(true))
+            .count()
+            .get_result::<i64>(conn)
+            .await?;
+
+        if primary_count > 0 {
+            return Ok(None); // User already has a primary email
+        }
+
+        self.insert(conn).await.map(Some)
     }
 
-    pub async fn insert_or_update(
+    // Inserts an email for the user, replacing the primary email if it exists.
+    pub async fn insert_or_update_primary(
         &self,
         conn: &mut AsyncPgConnection,
-    ) -> QueryResult<SecretString> {
-        diesel::insert_into(emails::table)
-            .values(self)
-            .on_conflict(emails::user_id)
-            .do_update()
-            .set(self)
-            .returning(emails::token)
-            .get_result::<String>(conn)
-            .await
-            .map(Into::into)
+    ) -> QueryResult<Email> {
+        // Attempt to update an existing primary email
+        let updated_email = diesel::update(
+            emails::table
+                .filter(emails::user_id.eq(self.user_id))
+                .filter(emails::primary.eq(true)),
+        )
+        .set((
+            emails::email.eq(self.email),
+            emails::verified.eq(self.verified),
+        ))
+        .returning(Email::as_returning())
+        .get_result(conn)
+        .await
+        .optional()?;
+
+        if let Some(email) = updated_email {
+            Ok(email)
+        } else {
+            // Otherwise, insert a new email
+            self.insert(conn).await
```

Comment:
probably not worth the extra complexity fixing it, but worth mentioning: if `insert_or_update_primary()` is called on a `NewEmail`  with `primary: false`, then this code path would not add it as a primary email. the only call site of this method is currently setting `primary: true`, but I'm wondering whether we should at least `error!(...)` log that case for now.